recipes/dev/async_grpo.py

```diff
@@ -1,114 +1,46 @@
-   This could be because
-    a. I turned compile off (it seems to be broken with torch 2.5.1 that vllm requires)
-    b. [FIXED] I turned activation checkpointing off (there wasn't an explcit reason for this it just got omitted accidentally)
-2. I have an assert that num_vllm_workers == 1 and vllm_tp_size == 1 for now. There's no real reason for this,
-   I expect the code to mostly generalize, just need to go over parts where I might have hardcoded this assumption of 1
-   vllm worker.
-3. Epochs is definitely not handled correctly right now :P
-4. There's many FIXMEs peppered through this script
-
-The run command is
-
-    python runnable_recipe_ray_vllm_weight_sync.py --config ../recipes/configs/dev/qwen3B_full_grpo.yaml
-
-
-"""
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
 
 import functools
 import os
 import time
-from functools import partial
-from typing import Any, Dict, List, Optional, Union
 
-from warnings import warn
+from typing import Any, Dict
 
 import ray
+
 import torch
 import torch.distributed
-import torch.nn as nn
-import torchtune.training as training
 
-from omegaconf import DictConfig, ListConfig, OmegaConf
-from ray.util.placement_group import placement_group
+from omegaconf import DictConfig, OmegaConf
 
-from ray.util.queue import Full as QueueFull, Queue
-from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+from ray.util.queue import Queue
 
-from readerwriterlock import rwlock
-from tensordict import TensorClass, TensorDict
+from tensordict import TensorDict, TensorDictBase
+from tensordict.utils import expand_as_right
 
-from torch.optim import Optimizer
-
-from torchdata.stateful_dataloader import StatefulDataLoader
-from torchdata.stateful_dataloader.sampler import StatefulDistributedSampler
-from torchrl.collectors import (
-    SyncDataCollector,
-    WeightUpdateReceiverBase,
-    WeightUpdateSenderBase,
-)
+from torchrl.collectors import WeightUpdateReceiverBase
 from torchrl.data import LazyStackStorage, RayReplayBuffer
-from torchtune import config, generation, modules, rlhf, utils
-from torchtune.dev.grpo.rewards import batched_rewards
-from torchtune.dev.grpo.types import GRPOStats, GRPOTrajectory
+from torchtune import config, utils
 from torchtune.dev.rl.datatypes import RequestOutput, Trajectory
-from torchtune.dev.rl.utils import stateless_init_process_group
 from torchtune.dev.rl.workers import (
     MetricLoggerActor,
     PyTorchActorModel,
     RefActor,
     SyncLLMCollector,
     VLLMParameterServer,
 )
-from torchtune.models.qwen2._convert_weights import qwen2_tune_to_hf
-
-from torchtune.training import DummyProfiler, PROFILER_KEY
+from torchtune.recipe_interfaces import OrchestrationRecipeInterface
+from vllm import SamplingParams
```

Comment:
Is vllm going to be a core dependency of tune?
Do we want an import at the root of vllm components?

Usually for recipe I think it's fine, since doing `from torchtune.foo import bar` should not touch the recipe, but if it does we must ensure that this doesn't raise an ImportError

Reply:
Yeah I definitely don't think we want vLLM (or Ray) as core dependencies in tune. E.g. just `pip install vllm` brings all these goodies with it: https://gist.github.com/ebsmothers/9f55c6ab1970cacecfef0828f91d3285. We should figure out a way to better decouple recipe environments, but until then maybe we can just gate?

Reply:
Ray is definitely just for RL but I think there's a good chance that vLLM ends up being needed in many places and recipes. I think the lowest effort way to support this for now is to treat this like bitsandbytes and in the config for the recipe include "This recipe requires `pip install vllm` and `pip install ray` to run." But we should figure out a better strategy.

Reply:
Let's just do an `[async_rl]` optional dep. On it.